README.md
# Plugin for GLPI

![Flyve MDM banner](https://user-images.githubusercontent.com/663460/26935464-54267e9c-4c6c-11e7-86df-8cfa6658133e.png)

[![License](https://img.shields.io/github/license/flyve-mdm/glpi-plugin.svg?&label=License)](https://github.com/flyve-mdm/glpi-plugin/blob/develop/LICENSE.md)
[![Follow twitter](https://img.shields.io/twitter/follow/FlyveMDM.svg?style=social&label=Twitter&style=flat-square)](https://twitter.com/FlyveMDM)
[![Telegram Group](https://img.shields.io/badge/Telegram-Group-blue.svg)](https://t.me/flyvemdm)
[![IRC Chat](https://img.shields.io/badge/IRC-%23flyvemdm-green.svg)](http://webchat.freenode.net/?channels=flyve-mdm)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-yellow.svg)](https://conventionalcommits.org)
[![Greenkeeper badge](https://badges.greenkeeper.io/flyve-mdm/glpi-plugin.svg)](https://greenkeeper.io/)
[![GitHub release](https://img.shields.io/github/release/flyve-mdm/glpi-plugin.svg)](https://github.com/flyve-mdm/glpi-plugin/releases)
[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)

Flyve MDM is a mobile device management software that enables you to secure and manage all the mobile devices of your business via a unique web-based console (MDM).

To get started, check out [Flyve MDM Website](https://flyve-mdm.com/)!

## Table of contents

* [Synopsis](#synopsis)
* [Build status](#build-status)
* [Compatibility Matrix](#compatibility-matrix)
* [Installation](#installation)
* [Documentation](#documentation)
* [Versioning](#versioning)
* [Contribute](#contribute)
* [Contact](#contact)
* [Professional Services](#professional-services)
* [Copying](#copying)

## Synopsis

This plugin for GLPI is a subproject of Flyve MDM.

It allows the management of mobile devices of your IT infrastructure, providing security policies, files and applications download and many more features.

It provides a stand alone web based administration and an interface integrated in GLPI. The later one may be the preferred by those who use GLPI for asset management and ticketing.

### Docker Environment

We created a docker environment for Flyve MDM, allowing you to have a development instance ready to work on, to know where to start, just follow our [documentation](http://flyve.org/docker-environment/howtos/).

## Build Status

| **LTS** | **Bleeding Edge** |
|:---:|:---:|
| [![Build Status](https://travis-ci.com/flyve-mdm/glpi-plugin.svg?branch=master)](https://travis-ci.com/flyve-mdm/glpi-plugin) | [![Build Status](https://travis-ci.com/flyve-mdm/glpi-plugin.svg?branch=develop)](https://travis-ci.com/flyve-mdm/glpi-plugin) |

## Compatibility Matrix

|GLPI|9.1.x|9.2.1+|
|:---|:---:|:---:|
|Flyve MDM|1.x|2.0.0|
|FusionInventory|9.1+1.0|9.2+2.0|
|Web MDM Dashboard|-|2.0.0|
|Android MDM Agent|-|2.0.0|

## Installation

Installation instructions are available on our documentation, see our [Installation section](http://flyvemdm-doc.readthedocs.io/en/latest/installation/index.html).

## Documentation

Checkout our [Documentation](http://flyvemdm-doc.readthedocs.io/en/latest/) hosted on Read the Docs.

We also share long-form content about the project on the website, check our [Development](http://flyve.org/glpi-plugin/) section.

## Versioning

In order to provide transparency on our release cycle and to maintain backward compatibility, Flyve MDM is maintained under [the Semantic Versioning guidelines](http://semver.org/). We are committed to following and complying with the rules, the best we can.

See [the tags section of our GitHub project](http://github.com/flyve-mdm/glpi-plugin/tags) for changelogs for each release version of Flyve MDM. Release announcement posts on [the official Teclib' blog](http://www.teclib-edition.com/en/communities/blog-posts/) contain summaries of the most noteworthy changes made in each release.

## Contribute

Want to file a bug, contribute some code, or improve documentation? Excellent! Read up on our
guidelines for [contributing](https://github.com/flyve-mdm/glpi-plugin/blob/develop/CONTRIBUTING.md) and then check out one of our issues in the [Issues Dashboard](https://github.com/flyve-mdm/glpi-plugin/issues).

If you are not a developer you may contribute the project by [translating it](https://www.transifex.com/flyve-mdm/public/), by [submitting documentation](https://github.com/flyve-mdm/glpi-plugin/) or by testing the product.

## Contact

For notices about major changes and general discussion of Flyve MDM development, subscribe to the [/r/FlyveMDM](http://www.reddit.com/r/FlyveMDM) subreddit.
You can also chat with us via IRC in [#flyve-mdm on freenode](http://webchat.freenode.net/?channels=flyve-mdm) or [@flyvemdm on Telegram](https://t.me/flyvemdm).
Ping me @btry in the IRC chatroom if you get stuck.

## Professional Services

The Flyve MDM and GLPI Network services are available through our [Partner's Network](http://www.teclib-edition.com/en/partners/). We provide special training, bug fixes with editor subscription, contributions for new features, and more.

Obtain a personalized service experience, associated with benefits and opportunities.

## Copying

* **Name**: [Flyve MDM](https://flyve-mdm.com/) is a registered trademark of [Teclib'](http://www.teclib-edition.com/en/).
* **Code**: you can redistribute it and/or modify it under the terms of the GNU AGPL v3 license ([AGPLv3](https://www.gnu.org/licenses/agpl-3.0.en.html)).
* **Documentation**: released under Attribution 4.0 International ([CC BY 4.0](https://creativecommons.org/licenses/by/4.0/)).


CHANGELOG.md
# [2.0.0](https://github.com/flyve-mdm/glpi-plugin/compare/v1.1.1...v2.0.0) (2018-12-21)


### Bug Fixes

* code style, drop compatibility with PHP 5.5 and official GLPI 9.1.x ([9c7c826](https://github.com/flyve-mdm/glpi-plugin/commit/9c7c826))
* **agent:** add missing input when enrol an agent ([88f9440](https://github.com/flyve-mdm/glpi-plugin/commit/88f9440))
* **agent:** better display of online status in history ([3623120](https://github.com/flyve-mdm/glpi-plugin/commit/3623120))
* **agent:** compatibility workaround for GLPI 9.3.0 ([d627bb2](https://github.com/flyve-mdm/glpi-plugin/commit/d627bb2))
* **agent:** description for mdm types ([dce5dcd](https://github.com/flyve-mdm/glpi-plugin/commit/dce5dcd))
* **agent:** find FI agent by name on enrolment ([727a8ae](https://github.com/flyve-mdm/glpi-plugin/commit/727a8ae))
* **agent:** fix paging ([d57df04](https://github.com/flyve-mdm/glpi-plugin/commit/d57df04))
* **agent:** fix typo in tests ([39d1200](https://github.com/flyve-mdm/glpi-plugin/commit/39d1200))
* **agent:** fixed agent account name on enrollment process. ([dc48776](https://github.com/flyve-mdm/glpi-plugin/commit/dc48776))
* **agent:** fixed XML inventory request, it must be on base64 ([31018f9](https://github.com/flyve-mdm/glpi-plugin/commit/31018f9))
* **agent:** getting an agent by topic failed ([d3e0e12](https://github.com/flyve-mdm/glpi-plugin/commit/d3e0e12))
* **agent:** machine account creation ([1f8e11f](https://github.com/flyve-mdm/glpi-plugin/commit/1f8e11f))
* **agent:** non conform definition of booleans ([ce5947f](https://github.com/flyve-mdm/glpi-plugin/commit/ce5947f))
* **agent:** redirect output from FI messages on enrollment ([67e5aa4](https://github.com/flyve-mdm/glpi-plugin/commit/67e5aa4))
* **agent:** removed action of delete computer on deletion ([cd01ad2](https://github.com/flyve-mdm/glpi-plugin/commit/cd01ad2))
* **agent:** removed base64 encoding for XML inventory ([89bc793](https://github.com/flyve-mdm/glpi-plugin/commit/89bc793))
* **agent:** reset device enroll status on delete (issue [#32](https://github.com/flyve-mdm/glpi-plugin/issues/32)) ([dd908fd](https://github.com/flyve-mdm/glpi-plugin/commit/dd908fd))
* **agent:** save system permission bit ([afb4938](https://github.com/flyve-mdm/glpi-plugin/commit/afb4938))
* **agent:** some mqtt topics not cleaned up ([73dbc94](https://github.com/flyve-mdm/glpi-plugin/commit/73dbc94)), closes [#285](https://github.com/flyve-mdm/glpi-plugin/issues/285)
* **agent:** sql error ([7692c46](https://github.com/flyve-mdm/glpi-plugin/commit/7692c46)), closes [#710](https://github.com/flyve-mdm/glpi-plugin/issues/710)
* **agent:** type of is_online is specific ([b44b5fa](https://github.com/flyve-mdm/glpi-plugin/commit/b44b5fa))
* **agent:** update checks for XML format on enrollment ([6e95d99](https://github.com/flyve-mdm/glpi-plugin/commit/6e95d99))
* **agent:** update search options ([9096104](https://github.com/flyve-mdm/glpi-plugin/commit/9096104)), closes [#179](https://github.com/flyve-mdm/glpi-plugin/issues/179)
* **agent:** update search options ([d0da4b3](https://github.com/flyve-mdm/glpi-plugin/commit/d0da4b3)), closes [#179](https://github.com/flyve-mdm/glpi-plugin/issues/179)
* **agent:** update the android app package id ([b7b0fa1](https://github.com/flyve-mdm/glpi-plugin/commit/b7b0fa1))
* **agent:** updated code for not register invalid user agents ([0ff79c1](https://github.com/flyve-mdm/glpi-plugin/commit/0ff79c1))
* **agent:** updated code for reload FI plugin ([8f6b5d9](https://github.com/flyve-mdm/glpi-plugin/commit/8f6b5d9))
* **agent:** wrong input variables assignment ([da36c3d](https://github.com/flyve-mdm/glpi-plugin/commit/da36c3d))
* **agent:** wrong user profile in enrollment ([a7f87a7](https://github.com/flyve-mdm/glpi-plugin/commit/a7f87a7))
* **agent:** xml of inventory is now decoded ([a73a1ed](https://github.com/flyve-mdm/glpi-plugin/commit/a73a1ed))
* **api:** file upload with GLPI 9.2 ([48b097a](https://github.com/flyve-mdm/glpi-plugin/commit/48b097a))
* **api:** restrict file and package download to the API only ([b2b5bec](https://github.com/flyve-mdm/glpi-plugin/commit/b2b5bec))
* **composer:** errors in composer ([89e125a](https://github.com/flyve-mdm/glpi-plugin/commit/89e125a))
* **config:** field to upload CA certificate file ([ed74f00](https://github.com/flyve-mdm/glpi-plugin/commit/ed74f00))
* **config:** fix errors in wizard ([#263](https://github.com/flyve-mdm/glpi-plugin/issues/263)) ([5399fde](https://github.com/flyve-mdm/glpi-plugin/commit/5399fde)), closes [#262](https://github.com/flyve-mdm/glpi-plugin/issues/262)
* **deployapp:** disable mqtt notifications on not parsed apps ([4ebafa8](https://github.com/flyve-mdm/glpi-plugin/commit/4ebafa8))
* **deployfile:** allow different files at same destination (fixes [#231](https://github.com/flyve-mdm/glpi-plugin/issues/231)) ([8b4c907](https://github.com/flyve-mdm/glpi-plugin/commit/8b4c907))
* **documentation:** fix headers in php source files ([1e7321a](https://github.com/flyve-mdm/glpi-plugin/commit/1e7321a))
* **documentation:** update table names for mosquitto auth plugin ([f21dd46](https://github.com/flyve-mdm/glpi-plugin/commit/f21dd46))
* **enrollment:** ACL reset on mqtt user creation ([ed48eaa](https://github.com/flyve-mdm/glpi-plugin/commit/ed48eaa))
* **enrollment:** ensure invitation token is unique ([1afea60](https://github.com/flyve-mdm/glpi-plugin/commit/1afea60))
* **enrollment:** fix computer type affectation in enrollment process ([3306264](https://github.com/flyve-mdm/glpi-plugin/commit/3306264))
* **enrollment:** fix deprecated methods call ([12305d8](https://github.com/flyve-mdm/glpi-plugin/commit/12305d8))
* **enrollment:** invitation emails attachment missing ([8dda7b3](https://github.com/flyve-mdm/glpi-plugin/commit/8dda7b3))
* **enrollment:** re-add base64_encode ([3f271ea](https://github.com/flyve-mdm/glpi-plugin/commit/3f271ea))

setup.php
<?php
/**
 * LICENSE
 *
 * Copyright © 2016-2018 Teclib'
 * Copyright © 2010-2018 by the FusionInventory Development Team.
 *
 * This file is part of Flyve MDM Plugin for GLPI.
 *
 * Flyve MDM Plugin for GLPI is a subproject of Flyve MDM. Flyve MDM is a mobile
 * device management software.
 *
 * Flyve MDM Plugin for GLPI is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * Flyve MDM Plugin for GLPI is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License
 * along with Flyve MDM Plugin for GLPI. If not, see http://www.gnu.org/licenses/.
 * ------------------------------------------------------------------------------
 * @copyright Copyright © 2018 Teclib
 * @license   https://www.gnu.org/licenses/agpl.txt AGPLv3+
 * @link      https://github.com/flyve-mdm/glpi-plugin
 * @link      https://flyve-mdm.com/
 * ------------------------------------------------------------------------------
 */

// Version of the plugin
define('PLUGIN_FLYVEMDM_VERSION', '2.0.0');
// Schema version of this version
define('PLUGIN_FLYVEMDM_SCHEMA_VERSION', '2.0');
// is or is not an official release of the plugin
define('PLUGIN_FLYVEMDM_IS_OFFICIAL_RELEASE', true);
// Minimal GLPI version, inclusive
define('PLUGIN_FLYVEMDM_GLPI_MIN_VERSION', '9.2');
// Maximum GLPI version, exclusive
define('PLUGIN_FLYVEMDM_GLPI_MAX_VERSION', '9.4');

define('PLUGIN_FLYVEMDM_ROOT', GLPI_ROOT . '/plugins/flyvemdm');

define('PLUGIN_FLYVEMDM_AGENT_DOWNLOAD_URL', 'https://play.google.com/store/apps/details?id=org.flyve.mdm.agent');

define('PLUGIN_FLYVEMDM_AGENT_BETA_DOWNLOAD_URL', 'https://play.google.com/apps/testing/org.flyve.mdm.agent');

define('PLUGIN_FLYVEMDM_APPLE_DEVELOPER_ENTERPRISE_URL', 'https://developer.apple.com/programs/enterprise/');

define('PLUGIN_FLYVEMDM_DEEPLINK', 'http://flyve.org/deeplink/');

if (!defined('FLYVEMDM_CONFIG_PATH')) {
   define('FLYVEMDM_CONFIG_PATH', GLPI_PLUGIN_DOC_DIR . '/flyvemdm');
}

if (!defined('FLYVEMDM_CONFIG_CACERTMQTT')) {
   define('FLYVEMDM_CONFIG_CACERTMQTT', FLYVEMDM_CONFIG_PATH . '/CACert-mqtt.crt');
}

if (!defined('FLYVEMDM_PACKAGE_PATH')) {
   define('FLYVEMDM_PACKAGE_PATH', GLPI_PLUGIN_DOC_DIR . '/flyvemdm/package');
}

if (!defined('FLYVEMDM_FILE_PATH')) {
   define('FLYVEMDM_FILE_PATH', GLPI_PLUGIN_DOC_DIR . '/flyvemdm/file');
}

if (!defined('FLYVEMDM_INVENTORY_PATH')) {
   define('FLYVEMDM_INVENTORY_PATH', GLPI_PLUGIN_DOC_DIR . '/flyvemdm/inventory');
}

if (!defined('FLYVEMDM_TEMPLATE_CACHE_PATH')) {
   define('FLYVEMDM_TEMPLATE_CACHE_PATH', GLPI_PLUGIN_DOC_DIR . '/flyvemdm/cache');
}

// Init the hooks of the plugins -Needed
function plugin_init_flyvemdm() {
   global $PLUGIN_HOOKS, $CFG_GLPI;

   $PLUGIN_HOOKS['csrf_compliant']['flyvemdm'] = true;
   $PLUGIN_HOOKS['undiscloseConfigValue']['flyvemdm'] = [PluginFlyvemdmConfig::class,
                                                         'undiscloseConfigValue'];

   $plugin = new Plugin();

   $config = Config::getConfigurationValues('flyvemdm', ['version']);
   if (isset($config['version']) && $config['version'] != PLUGIN_FLYVEMDM_VERSION) {
      $plugin->getFromDBbyDir('flyvemdm');
      $plugin->update([
            'id'     => $plugin->getID(),
            'state'  => Plugin::NOTUPDATED
      ]);
   }

   if (!$plugin->getFromDBbyDir('flyvemdm')) {
      // nothing more to do at this moment
      return;
   }

   $state = $plugin->getField('state');
   if ($state != Plugin::NOTACTIVATED) {
      require_once(__DIR__ . '/vendor/autoload.php');
   }

   if ($state == Plugin::ACTIVATED) {
      if (!class_exists('GlpiLocalesExtension')) {
         require_once(__DIR__ . '/lib/GlpiLocalesExtension.php');
      }

      plugin_flyvemdm_registerClasses();
      plugin_flyvemdm_addHooks();

      Html::requireJs('charts');
      $PLUGIN_HOOKS['add_css']['flyvemdm'][] = "css/style.css";
      // Warning 'pluginflyvemdmmenu' MUST be lower case
      $CFG_GLPI['javascript']['admin']['pluginflyvemdmmenu']['Menu'] = ['charts'];

      if (strpos($_SERVER['REQUEST_URI'], "plugins/flyvemdm/front/agent.form.php") !== false) {
         $PLUGIN_HOOKS['add_css']['flyvemdm'][] = 'lib/leaflet-1.0.3/leaflet.css';
         $PLUGIN_HOOKS['add_javascript']['flyvemdm'][] = 'lib/leaflet-1.0.3/leaflet.js';
      }
   }
}

/**
 * Register classes
 */
function plugin_flyvemdm_registerClasses() {
   Plugin::registerClass(PluginFlyvemdmAgent::class);
   Plugin::registerClass(PluginFlyvemdmFleet::class);
   Plugin::registerClass(PluginFlyvemdmPolicy::class);
   Plugin::registerClass(PluginFlyvemdmTask::class);
   Plugin::registerClass(PluginFlyvemdmProfile::class,
         ['addtabon' => Profile::class]);
   Plugin::registerClass(PluginFlyvemdmGeolocation::class);
   Plugin::registerClass(PluginFlyvemdmPackage::class);
   Plugin::registerClass(PluginFlyvemdmFile::class);
   Plugin::registerClass(PluginFlyvemdmInvitation::class,
         ['notificationtemplates_types' => true, /* 'document_types' => true */]);
   Plugin::registerClass(PluginFlyvemdmEntityConfig::class,
         ['addtabon' => Entity::class]);

   // Dropdowns
   Plugin::registerClass(PluginFlyvemdmWellknownpath::class);
   Plugin::registerClass(PluginFlyvemdmPolicyCategory::class);
}

/**
 * Adds all hooks the plugin needs
 */
function plugin_flyvemdm_addHooks() {
   global $PLUGIN_HOOKS;

   $PLUGIN_HOOKS['change_profile']['flyvemdm'] = [PluginFlyvemdmProfile::class, 'changeProfile'];

   //if glpi is loaded
   if (Session::getLoginUserID()) {
      $PLUGIN_HOOKS['menu']['flyvemdm'] = true;
   }
   $PLUGIN_HOOKS['post_init']['flyvemdm'] = 'plugin_flyvemdm_postinit';

   if (Session::haveRight(PluginFlyvemdmProfile::$rightname, PluginFlyvemdmProfile::RIGHT_FLYVEMDM_USE)) {
      // Define menu entries
      $PLUGIN_HOOKS['menu_toadd']['flyvemdm'] = [
         'admin'  => 'PluginFlyvemdmMenu',
      ];
      $PLUGIN_HOOKS['autoinventory_information']['flyvemdm'] = [
         'Computer' => [PluginFlyvemdmAgent::class, 'displayTabContentForComputer'],
      ];
      $PLUGIN_HOOKS['config_page']['flyvemdm'] = 'front/config.form.php';
   }

   // Hooks for the plugin : objects inherited from GLPI or
   $PLUGIN_HOOKS['item_add']['flyvemdm']     = [
      Entity::class                    => 'plugin_flyvemdm_hook_entity_add',
   ];
   $PLUGIN_HOOKS['item_purge']['flyvemdm']   = [
      Entity::class                    => 'plugin_flyvemdm_hook_entity_purge',
      Computer::class                  => 'plugin_flyvemdm_hook_computer_purge',
   ];
   $PLUGIN_HOOKS['pre_item_purge']['flyvemdm']   = [
      PluginFlyvemdmInvitation::class => 'plugin_flyvemdm_hook_pre_invitation_purge',
      Document::class                 => [PluginFlyvemdmInvitation::class, 'hook_pre_document_purge'],
      Profile_User::class             => 'plugin_flyvemdm_hook_pre_profileuser_purge',
   ];

   // Notifications
   $PLUGIN_HOOKS['item_get_events']['flyvemdm'] = [];
   $PLUGIN_HOOKS['item_get_datas']['flyvemdm'] = [];

   $PLUGIN_HOOKS['item_get_events']['flyvemdm'][PluginFlyvemdmNotificationTargetInvitation::class] = [
      PluginFlyvemdmNotificationTargetInvitation::class, 'addEvents'
   ];
   $PLUGIN_HOOKS['item_get_datas']['flyvemdm'][PluginFlyvemdmNotificationTargetInvitation::class] = [
      PluginFlyvemdmNotificationTargetInvitation::class, 'getAdditionalDatasForTemplate'
   ];

   $PLUGIN_HOOKS['use_massive_action']['flyvemdm'] = 1;

   $PLUGIN_HOOKS['import_item']['flyvemdm'] = [Computer::class => ['Plugin']];
}

/**
 * Get the name and the version of the plugin - Needed
 * @return array
 */
function plugin_version_flyvemdm() {
   $author = '<a href="http://www.teclib.com">Teclib</a>';
   $requirements = [
      'name'           => __s('Flyve Mobile Device Management', 'flyvemdm'),
      'version'        => PLUGIN_FLYVEMDM_VERSION,
      'author'         => $author,
      'license'        => 'AGPL-3.0-or-later',
      'homepage'       => 'https://flyve-mdm.com/',
      'minGlpiVersion' => PLUGIN_FLYVEMDM_GLPI_MIN_VERSION,
      'requirements'   => [
         'glpi' => [
            'min' => PLUGIN_FLYVEMDM_GLPI_MIN_VERSION,
            'plugins'   => [
               'fusioninventory',
            ],
            'params' => [],
         ],
         'php' => [
            'exts'   => [
               'OpenSSL'   => [
                  'required'  => true,
               ],
               'sockets'   => [
                  'required'  => true,
                  'function'  => 'socket_create'
               ],
               'zip'       => [
                  'required'  => true,
               ],
               'gd'        => [
                  'required'  => true
               ]
            ]
         ]
      ]
    ];

   if (PLUGIN_FLYVEMDM_IS_OFFICIAL_RELEASE) {
      // This is not a development version
      $requirements['requirements']['glpi']['max'] = PLUGIN_FLYVEMDM_GLPI_MAX_VERSION;
   }
   return $requirements;
}

/**
 * Check pre-requisites before install
 * OPTIONNAL, but recommanded
 *
 * @return boolean
 */
function plugin_flyvemdm_check_prerequisites() {
   $prerequisitesSuccess = true;

   if (!is_readable(__DIR__ . '/vendor/autoload.php') || !is_file(__DIR__ . '/vendor/autoload.php')) {
      echo "Run composer install --no-dev in the plugin directory<br>";
      $prerequisitesSuccess = false;
   }

   if (version_compare(GLPI_VERSION, PLUGIN_FLYVEMDM_GLPI_MIN_VERSION, 'lt')
       || PLUGIN_FLYVEMDM_IS_OFFICIAL_RELEASE && version_compare(GLPI_VERSION, PLUGIN_FLYVEMDM_GLPI_MAX_VERSION, 'ge')) {
      echo "This plugin requires GLPi >= " . PLUGIN_FLYVEMDM_GLPI_MIN_VERSION . " and GLPI < " . PLUGIN_FLYVEMDM_GLPI_MAX_VERSION . "<br>";
      $prerequisitesSuccess = false;
   }

   return $prerequisitesSuccess;
}

/**
 * Uninstall process for plugin : need to return true if succeeded : may display messages or add to message after redirect
 * @param boolean $verbose Whether to display message on failure. Defaults to false
 * @return bool
 */
function plugin_flyvemdm_check_config($verbose = false) {
   return true;
}

/**
 * @return Twig_Environment
 */
function plugin_flyvemdm_getTemplateEngine() {
   $loader = new Twig_Loader_Filesystem(__DIR__ . '/tpl');
   $twig =  new Twig_Environment($loader, [
         'cache'        => FLYVEMDM_TEMPLATE_CACHE_PATH,
         'auto_reload'  => ($_SESSION['glpi_use_mode'] == Session::DEBUG_MODE),
   ]);
   $twig->addExtension(new GlpiLocalesExtension());
   return $twig;
}

/**
 * center all columns of plugin
 * @param string $itemtype
 * @param integer $ID
 * @param mixed $data
 * @param integer $num
 * @return string
 */
function plugin_flyvemdm_displayConfigItem($itemtype, $ID, $data, $num) {
   return "align='center'";
}

/**
 * Show the last SQL error, logs its backtrace and dies
 * @param Migration $migration
 */
function plugin_flyvemdm_upgrade_error(Migration $migration) {
   global $DB;

   $error = $DB->error();
   $migration->log($error . "\n" . Toolbox::backtrace(false, '', ['Toolbox::backtrace()']), false);
   die($error . "<br><br> Please, check migration log");
}
